### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Groupoid

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #2230 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2230
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2230~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Groupoid.lean [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+See also `CategoryTheory.Core` for the groupoid of isomorphisms in a category.
+-/
+
+
+namespace CategoryTheory
+
+universe v v₂ u u₂
+
+-- morphism levels before object levels. See note [CategoryTheory universes].
+/-- A `groupoid` is a category such that all morphisms are isomorphisms. -/
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Groupoid.lean:43]:
```suggestion
/-- A `Groupoid` is a category such that all morphisms are isomorphisms. -/
```

File: Mathlib/CategoryTheory/Groupoid.lean [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+instance (priority := 100) IsIso.of_groupoid (f : X ⟶ Y) : IsIso f :=
+  ⟨⟨Groupoid.inv f, Groupoid.comp_inv f, Groupoid.inv_comp f⟩⟩
+#align category_theory.is_iso.of_groupoid CategoryTheory.IsIso.of_groupoid
+
+@[simp]
+theorem Groupoid.inv_eq_inv (f : X ⟶ Y) : Groupoid.inv f = CategoryTheory.inv f :=
+  IsIso.eq_inv_of_hom_inv_id <| Groupoid.comp_inv f
+#align category_theory.groupoid.inv_eq_inv CategoryTheory.Groupoid.inv_eq_inv
+
+/-- `groupoid.inv` is involutive. -/
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Groupoid.lean:81]:
```suggestion
/-- `Groupoid.inv` is involutive. -/
```

File: Mathlib/CategoryTheory/Groupoid.lean [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import Mathlib.CategoryTheory.FullSubcategory
+import Mathlib.CategoryTheory.Products.Basic
+import Mathlib.CategoryTheory.Pi.Basic
+import Mathlib.CategoryTheory.Category.Basic
+import Mathlib.Combinatorics.Quiver.ConnectedComponent
+
+/-!
+# Groupoids
+
+We define `groupoid` as a typeclass extending `category`,
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Groupoid.lean:20]:
```suggestion
We define `Groupoid` as a typeclass extending `category`,
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**